### PR TITLE
cincinnati: expose dead-end release information via MOTD

### PR DESCRIPTION
**Pull request by sohankunkerkar**
After checking whether the current Cincinnati node is a dead-end or
not, write out a new MOTD file at `/run/motd.d/deadend.motd`
reflecting the dead-end status. If the current release is a deadend,
a message is printed in the MOTD containing the reason for the
release being a deadend. An info message is also logged. If not a
dead-end release a blank MOTD is written and no message is logged.

To make the final write atomic, a tempfile is written within the
`/run/motd.d/` directory, and renamed to the location
of the complete MOTD to display.

In this case, "pkexec" utility is used to allow Zincati (which runs as
the `zincati` sysuser) to modify root-owned `/run/motd.d/` in order to
write dead-end release information into a MOTD file, and the polkit rule
will ensure that that utility runs only when its invoked by a `zincati-deadend`
binary and the corresponding user is `zincati`.

Resolves: #90
Co-authored by: Robert Fairley (rob.t.fairley@gmail.com)

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -126,3 +131,34 @@ fn run_agent() -> failure::Fallible<()> {
 
     Ok(())
 }
+
+// Deadend subcommand entry point
```
Comment (lucab) [inline src/main.rs:135]: Nit: docstrings require three slashes (`///`) and usually end with a dot.

File: src/main.rs [hunk 2]
```diff
@@ -126,3 +131,34 @@ fn run_agent() -> failure::Fallible<()> {
...
+    if let Some(reason) = reason {
+        writeln!(
+            f,
+            "This release is a dead-end and won't auto-update: {}",
+            reason
+        )
+        .with_context(|e| format!("failed to write MOTD: {}", e))?;
+    }
+
+    f.persist("/run/motd.d/deadend.motd")
```
Comment (lucab) [inline src/main.rs:161]: It would be good to give this a numerical sorting and the application prefix, e.g. `30-zincati-deadend.motd`.

**Comment by lucab** [comment]
Thanks for the PR, I've left a few comment in review. The approach overall looks good, and indeed less messy than the symlinks chain. I think the biggest gaps in this WIP right now are 1) how to unset a previous deadend and 2) conditional logic in order to only act on state changes.

/cc @travier @jlebon @rfairley in case they want to have a look too.

**Comment by travier** [comment]
Agree with @lucab here. Nice catch for the argv1.

File: dist/polkit-1/rules.d/zincati.rules [hunk 3]
```diff
@@ -5,4 +5,12 @@ polkit.addRule(function(action, subject) {
         subject.user == "zincati") {
         return polkit.Result.YES;
     }
-})
+});
+
+// Allow Zincati to write dead-end release information into a MOTD file.
+polkit.addRule(function(action, subject) {
+    if (action.id == "org.freedesktop.policykit.exec" &&
+        action.lookup("program") == "/usr/libexec/zincati-deadend" &&  subject.user == "zincati") {
```
[the remark this answers is not in the document]
Comment (cgwalters) [reply dist/polkit-1/rules.d/zincati.rules:13]: Yeah, seems likely.

File: src/cli/mod.rs [hunk 4]
```diff
@@ -29,7 +29,11 @@ impl CliOptions {
 
 /// CLI sub-commands.
 #[derive(Debug, StructOpt)]
+#[structopt(rename_all = "kebab-case")]
```
[the remark this answers is not in the document]
Comment (cgwalters) [reply src/cli/mod.rs:32]: Ah heh.  I advised the opposite and ended up doing https://github.com/coreos/zincati/pull/368

**Comment by cgwalters** [comment]
Not to completely derail this but I did think of a potentially much simpler solution that doesn't involve polkit:
`ExecStartPre=+/usr/bin/setfacl -m u:zincati:rwx /run/motd.d`
Then zincati can just write to `/run/motd.d`.  (Now yes it could unlink or modify other motd files but eh)

**Comment by rfairley** [comment]
:+1: I like the idea of keeping the final MOTD within `/run/motd.d` - avoids the need to have a corresponding SELinux rule added as the implementation in #276 needs. Note if this route is taken, this commit https://github.com/fedora-selinux/selinux-policy/pull/398/commits/760811d670d0399c38df64afe9f83ded4f96d5ef should be reverted.

One concern on the approach I have is: could the temporary `deadend.motd.partial` file be read by pam_motd callers, e.g. sshd, sometime before the final rename to `deadend.motd`, causing partial output to be displayed? It's unlikely that logins will hit this, but if this can happen it may take away from the original benefit from using the tmpfile+rename pattern.

If this is the case, one way around the issue may be to modify `pam_motd` to only display filenames with `.motd` suffix (so a `.motd.partial` file would never be displayed) - see https://github.com/linux-pam/linux-pam/blob/e21fd1f344a67844925ab9a06b0f54299c94b56c/modules/pam_motd/pam_motd.c#L171. `login` does filter MOTDs by `.motd` suffix (https://github.com/karelzak/util-linux/commit/5a528e2c6ff9735266fc2607c359e925b074bf2c#diff-77a7801e7ac65d85ef624043d0dd2647R250) .

File: src/cincinnati/mod.rs [hunk 5]
```diff
@@ -55,6 +56,7 @@ lazy_static::lazy_static! {
         "Total number of errors while checking for updates.",
         &["kind"]
     ).unwrap();
+    static ref DEADEND_RELEASE_EXISTS : AtomicBool = AtomicBool::new(false);
```
Comment (lucab) [inline src/cincinnati/mod.rs:59]: As I mentioned in my [previous review](https://github.com/coreos/zincati/pull/367#discussion_r498668206), this is a [tristate](https://en.wikipedia.org/wiki/Three-valued_logic) (`unknown` / `true` / `false`), not a boolean.

In particular, when the service starts it doesn't know its deadend state (i.e. it starts in `unknown` mode) until the first fetch. This currently assumes the service is always starting from `false`, which is not guaranteed if the service was running before with a `true` state and then got restarted.

You can model this by wrapping an [AtomicU8](https://doc.rust-lang.org/std/sync/atomic/struct.AtomicU8.html) into a new object and exposing relevant set and query methods.

File: src/main.rs [hunk 6]
```diff
@@ -126,3 +131,38 @@ fn run_agent() -> failure::Fallible<()> {
...
+                "This release is a dead-end and won't auto-update: {}",
+                reason
+            )
+            .with_context(|e| format!("failed to write MOTD: {}", e))?;
+        }
+
+        f.persist("/run/motd.d/85-zincati-deadend.motd")
+            .with_context(|e| format!("failed to persist temporary MOTD file: {}", e))?;
+    } else {
+        std::fs::remove_file("/run/motd.d/85-zincati-deadend.motd")?;
```
Comment (lucab) [inline src/main.rs:165]: I think you'll need to special case the file not found case here, which is not an error condition in this case.

**Comment by lucab** [comment]
@cgwalters nice. It can even be described declaratively via `systemd-tmpfiles`. However it goes in the same direction of the previous symlinking approach, i.e. the unprivileged `zincati` user being able to freely change (in particular, symlink) global content. After @travier feedback, I think there was some consensus about trying to avoid that.

@rfairley good point. The tmpfile is a dotfile (`.deadend[...]`) under the assumption that hidden files won't be sourced by the MOTD assembling logic. I didn't check all the consumers to make sure they actually behave that way, but it may be worth patching them in that direction if it isn't the case already.

**Comment by sohankunkerkar** [comment]
@lucab Could you please take another look? 

> One concern on the approach I have is: could the temporary deadend.motd.partial file be read by pam_motd callers, e.g. sshd, sometime before the final rename to deadend.motd, causing partial output to be displayed? It's unlikely that logins will hit this, but if this can happen it may take away from the original benefit from using the tmpfile+rename pattern.

I'm not sure how to reproduce this issue locally (if at all it exists).

**Comment by sohankunkerkar** [comment]
> If this is the case, one way around the issue may be to modify pam_motd to only display filenames with .motd suffix (so a .motd.partial file would never be displayed) - see https://github.com/linux-pam/linux-pam/blob/e21fd1f344a67844925ab9a06b0f54299c94b56c/modules/pam_motd/pam_motd.c#L171. login does filter MOTDs by .motd suffix (karelzak/util-linux@5a528e2#diff-77a7801e7ac65d85ef624043d0dd2647R250) .

This sounds like a good idea. I can submit a patch upstream if it's a viable option and can mitigate the issue related to motd displaying only partial output.

File: src/cincinnati/mod.rs [hunk 7]
```diff
@@ -55,6 +57,26 @@ lazy_static::lazy_static! {
...
+#[derive(Debug, Clone)]
+pub enum DeadEndState {
+    Unknown = 0,
+    True = 1,
+    False = 2,
+}
+
+impl From<Arc<AtomicU8>> for DeadEndState {
+    fn from(x: Arc<AtomicU8>) -> Self {
+        match x.load(Ordering::Relaxed) {
```
Comment (lucab) [inline src/cincinnati/mod.rs:73]: Here and below, mind detailing why you went with `Relaxed` ordering?

File: src/main.rs [hunk 8]
```diff
@@ -7,7 +7,8 @@
 extern crate fail;
 #[macro_use]
 extern crate prometheus;
-
+#[macro_use]
+extern crate failure;
```
[the remark this answers is not in the document]
Comment (sohankunkerkar) [reply src/main.rs:11]: Agree, let me delete that line.

File: src/cincinnati/mod.rs (hunk 7 above)
Comment (sohankunkerkar) [reply src/cincinnati/mod.rs]: I think that was a bad choice

File: dist/polkit-1/rules.d/org.freedesktop.policykit.pkexec.run-zincati-deadend.policy [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+ "-//freedesktop//DTD PolicyKit Policy Configuration 1.0//EN"
+ "http://www.freedesktop.org/standards/PolicyKit/1/policyconfig.dtd">
+<policyconfig>
+  <action id="org.freedesktop.policykit.pkexec.run-zincati-deadend">
+    <description>Allow Zincati to write dead-end release information</description>
+    <defaults>
+      <allow_any>yes</allow_any>
+      <allow_inactive>yes</allow_inactive>
+      <allow_active>yes</allow_active>
+    </defaults>
```
Comment (travier) [inline dist/polkit-1/rules.d/org.freedesktop.policykit.pkexec.run-zincati-deadend.policy:12]: I think this should ideally be removed or at least set to no as in this current form this allows anyone to call the action.

File: dist/polkit-1/rules.d/org.freedesktop.policykit.pkexec.run-zincati-deadend.policy [hunk 10]
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE policyconfig PUBLIC
+ "-//freedesktop//DTD PolicyKit Policy Configuration 1.0//EN"
+ "http://www.freedesktop.org/standards/PolicyKit/1/policyconfig.dtd">
+<policyconfig>
+  <action id="org.freedesktop.policykit.pkexec.run-zincati-deadend">
```
Comment (travier) [inline dist/polkit-1/rules.d/org.freedesktop.policykit.pkexec.run-zincati-deadend.policy:6]:
```suggestion
  <action id="org.coreos.zincati.deadend">
```

File: dist/polkit-1/rules.d/zincati.rules [hunk 11]
```diff
@@ -5,4 +5,12 @@ polkit.addRule(function(action, subject) {
         subject.user == "zincati") {
         return polkit.Result.YES;
     }
-})
+});
+
+// Allow Zincati to write dead-end release information into a MOTD file.
+polkit.addRule(function(action, subject) {
+    if (action.id == "org.freedesktop.policykit.pkexec.run-zincati-deadend" &&
```
Comment (travier) [inline dist/polkit-1/rules.d/zincati.rules:12]:
```suggestion
    if (action.id == "org.coreos.zincati.deadend" &&  
```

File: src/main.rs [hunk 12]
```diff
@@ -126,3 +131,40 @@ fn run_agent() -> failure::Fallible<()> {
...
+            .prefix(".deadend.")
+            .suffix(".motd.partial")
+            // Create the tempfile in the same directory as the final MOTD,
+            // to ensure proper SELinux labels are applied to the tempfile
+            // before renaming.
+            .tempfile_in("/run/motd.d")
+            .with_context(|e| format!("failed to create temporary MOTD file: {}", e))?;
+        // Set correct permissions of the temporary file, before moving to
+        // the destination (`tempfile` creates files with mode 0600).
+        fs::set_permissions(f.path(), Permissions::from_mode(0o664))
```
Comment (travier) [inline src/main.rs:150]:
```suggestion
        fs::set_permissions(f.path(), Permissions::from_mode(0o644))
```

Any particular not to use `644` here?

**Review by travier — COMMENTED** [review]
Some changes requested but LGTM overall. I'm not sure about the polkit action changes and we will have to double check that this does not allow anyone from calling the action.

File: dist/polkit-1/rules.d/org.freedesktop.policykit.pkexec.run-zincati-deadend.policy (hunk 9 above)
Comment (travier) [reply dist/polkit-1/rules.d/org.freedesktop.policykit.pkexec.run-zincati-deadend.policy:12]: From https://wiki.archlinux.org/index.php/Polkit#Actions

File: src/main.rs (hunk 12 above)
Comment (sohankunkerkar) [reply src/main.rs:150]: hmm... I just carried this change from the (previous)[https://github.com/coreos/zincati/pull/276] PR, so I don't have any strict opinion here.

File: dist/polkit-1/rules.d/org.freedesktop.policykit.pkexec.run-zincati-deadend.policy (hunk 9 above)
Comment (sohankunkerkar) [reply dist/polkit-1/rules.d/org.freedesktop.policykit.pkexec.run-zincati-deadend.policy:12]: Yeah, makes sense. I was under the impression that this action will get called only for the `zincati` user when invoked by the `zincati.rules`(https://github.com/coreos/zincati/pull/367/files#diff-fb15afa4bf5dbf6d827a99f2d5d1af3eR13)

Comment (sohankunkerkar) [reply dist/polkit-1/rules.d/org.freedesktop.policykit.pkexec.run-zincati-deadend.policy:12]: WDYT about this change?
```
                        <allow_any>no</allow_any>
                        <allow_inactive>no</allow_inactive>
                        <allow_active>auth_admin_keep</allow_active>
```

Comment (travier) [reply dist/polkit-1/rules.d/org.freedesktop.policykit.pkexec.run-zincati-deadend.policy:12]: I think it should always be no for all entries here as this will never be called interactively (zincati is a daemon) and we only want to allow the case covered in the additional polkit rule we add.

File: dist/polkit-1/rules.d/org.freedesktop.policykit.pkexec.run-zincati-deadend.policy [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+    <description>Allow Zincati to write dead-end release information</description>
```
Comment (travier) [inline dist/polkit-1/rules.d/org.freedesktop.policykit.pkexec.run-zincati-deadend.policy:7]:
```suggestion
    <description>Write dead-end release information as an MOTD fragment via Zincati</description>
```

File: src/main.rs (hunk 12 above)
Comment (lucab) [reply src/main.rs:150]: @sohankunkerkar is the resulting file owned by `root:root`? If so, this is likely not going to matter but `0644` should be enough.

Comment (sohankunkerkar) [reply src/main.rs:150]: Yeah, it's owned by `root:root`
```
[core@localhost ~]$ ls -ltr /run/motd.d/85-zincati-deadend.motd 
-rw-rw-r--. 1 root root 109 Oct  9 15:43 /run/motd.d/85-zincati-deadend.motd

```

**Comment by lucab** [comment]
Self-notes:
 * the global static `DEADEND_STATE` is not nice to have but carving it out requires a larger refactoring. It can be tackled later on after merging this feature PR first.
 * shelling out to pkexec is a synchronous operation which should not performed in the middle of the async chain. It can indefinitely block and stall the executor. In the happy path this should not be a big concern, so let's tackle this afterwards.
 * in the hot-path of update checking we don't really care if deadend-motd refresh failed for some reason, so a fire-and-forget approach may be enough. Or at least a timeout in order to make sure the core loop keeps going.

All points above can be likely solved by moving this logic to its own `SyncArbiter`.

**Comment by lucab** [comment]
@sohankunkerkar can you please rebase on latest `master` and drop the WIP?

File: Cargo.toml [hunk 14]
```diff
@@ -10,33 +10,34 @@ repository = "https://github.com/coreos/zincati"
 edition = "2018"
 
 [dependencies]
-actix = "0.10"
```
Comment (travier) [inline Cargo.toml:13]: Looks like a bad rebase here.

Comment (sohankunkerkar) [reply Cargo.toml]: @travier Just rectified it. Could you please take another look?

**Review by travier — CHANGES_REQUESTED** [review]
Two minor nits, otherwise LGTM.

**Comment by sohankunkerkar** [comment]
> Self-notes:
> 
>     * the global static `DEADEND_STATE` is not nice to have but carving it out requires a larger refactoring. It can be tackled later on after merging this feature PR first.
> 
>     * shelling out to pkexec is a synchronous operation which should not performed in the middle of the async chain. It can indefinitely block and stall the executor. In the happy path this should not be a big concern, so let's tackle this afterwards.
> 
>     * in the hot-path of update checking we don't really care if deadend-motd refresh failed for some reason, so a fire-and-forget approach may be enough. Or at least a timeout in order to make sure the core loop keeps going.
> 
> 
> All points above can be likely solved by moving this logic to its own `SyncArbiter`.

@lucab Should I capture this point in Jira? It would be helpful for tracking purposes.

**Review by travier — APPROVED** [review]

**Comment by travier** [comment]
OK, how do we test that? Should we make a kola test?

**Comment by lucab** [comment]
@sohankunkerkar no need to right now, I'll explore a bit the option later and then if I find an arrangement that seems to fit better I'll open a ticket here.

**Comment by travier** [comment]
We also need to make a test to make sure only the Zincati user gets access to this action.

**Comment by sohankunkerkar** [comment]
> OK, how do we test that? Should we make a kola test?
Yeah, good point. The kola test would definitely cover the corner cases.

**Comment by lucab** [comment]
@travier I checked the polkit config (only, without binary changes) on latest FCOS `stable` this way:
```
$ id
uid=1000(core) gid=1000(core) groups=1000(core),4(adm),10(wheel),16(sudo),190(systemd-journal) context=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023
$ pkexec /usr/libexec/zincati deadend
Error executing command as another user: Not authorized

$ id
uid=979(zincati) gid=979(zincati) groups=979(zincati) context=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023
$ pkexec /usr/libexec/zincati deadend
USAGE:
    zincati [FLAGS] <SUBCOMMAND>

$ id
uid=979(zincati) gid=979(zincati) groups=979(zincati) context=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023
$ pkexec /usr/libexec/zincati agent
==== AUTHENTICATING FOR org.freedesktop.policykit.exec ====
Authentication is needed to run `/usr/libexec/zincati' as the super user
```

I think it works as expected, do you have other things in mind to check?

**Review by lucab — CHANGES_REQUESTED** [review]

**Review by lucab — CHANGES_REQUESTED** [review]

**Review by lucab — APPROVED** [review]

**Comment by lucab** [comment]
LGTM, thanks for the multiple iterations towards this!

**Comment by travier** [comment]
Thanks @sohankunkerkar!
